### DISCUSSION TITLE
code navigation

### DISCUSSION THREAD

**Discussion by lesorubshayan** [vim/vim#14716] [Q&A]
grep
cscope
ctags

grep is overhead
cscope and ctags may not work(for go to definition)

is there any other way?

**Comment by lesorubshayan** [comment]
how to locate c functions(e.g. for 'f' command or other that i know the binding of)?

**Comment by lesorubshayan** [comment]
@zeertzjq

**Comment by EdwardMG** [comment]
Have a look at `:h tjump` and `:h tselect`. These two commands will show you a list of matches from the tags files when multiple match so you can just type a number from a list to select the right one. cscope has similar methods. Sometimes you'll be surprised what you learn when you see all the places reusing a symbol.

If you know that the definition is in the current scope or file `:h gd` and `:h gD` can be quite effective. The are some others like `:h [I` and `:h [d` and many more hiding under `[` that support various kinds of movement and display for C-like syntax or preprocessor definitions.

tags does not do a very sophisticated search, it just prioritize speed through a binary search on the tags file. It doesn't know anything about the location that you are jumping from, just the term that you pass it. If you have the desire for it, it's not too difficult to write a slower but more accurate search that leverages your domain knowledge of the codebase you are searching.

I don't know of many better methods built into vim, but vim is very friendly to the shell for whatever tools you want to leverage. To be sure, code search is difficult when many different things share the same symbol. Sometimes you need more information than just the word, or in a perfect world better conventions in the codebase you have to work in to avoid so much clashing or frequent jumping between files.

Best of luck, hope it helps you even a month later :)
